Content.Client/Medical/Cryogenics/BeakerBarChart.cs

```diff
@@ -0,0 +1,268 @@
+        Color color,
+        Color? textColor = null,
+        string? tooltip = null)
+    {
+        // If we can find an old entry we're allowed to update, update that one.
+        if (TryFindUpdateableEntry(uid, out var index))
+        {
+            _entries[index].TargetAmount = amount;
+            _entries[index].Tooltip = tooltip;
+            _entries[index].Label.Text = label;
+            _nextUpdateableEntry = index + 1;
+            return;
+        }
+
+        // Otherwise create a new entry.
+        if (amount <= 0)
+            return;
+
+        // If no text color is provided, use either white or black depending on how dark the background is.
+        textColor ??= (color.R + color.G + color.B < 1.5f ? Color.White : Color.Black);
+
+        var childLabel = new Label
+        {
+            Text = label,
+            ClipText = true,
+            FontColorOverride = textColor,
+            Margin = new Thickness(4, 0, 0, 0)
+        };
+        AddChild(childLabel);
+
+        _entries.Insert(
+            _nextUpdateableEntry,
+            new Entry(uid, childLabel)
+            {
+                WidthFraction = (_hasBeenDrawn ? 0 : amount / Capacity),
+                TargetAmount = amount,
+                Tooltip = tooltip,
+                Color = color
+            }
+        );
+
+        _nextUpdateableEntry += 1;
+    }
+
+    private bool TryFindUpdateableEntry(string uid, out int index)
+    {
+        for (int i = _nextUpdateableEntry; i < _entries.Count; i++)
+        {
+            if (_entries[i].Uid == uid)
+            {
+                index = i;
+                return true;
+            }
+        }
+
+        index = -1;
+        return false;
+    }
+
+    private IEnumerable<(Entry, float xMin, float xMax)> EntryRanges(float? pixelWidth = null)
+    {
+        float chartWidth = pixelWidth ?? PixelWidth;
+        var xStart = 0f;
+
+        foreach (var entry in _entries)
+        {
+            var entryWidth = entry.WidthFraction * chartWidth;
+            var xEnd = MathF.Min(xStart + entryWidth, chartWidth);
+
+            yield return (entry, xStart, xEnd);
+
+            xStart = xEnd;
+        }
+    }
+
+    private bool TryFindEntry(float x, [NotNullWhen(true)] out Entry? entry)
+    {
+        foreach (var (currentEntry, xMin, xMax) in EntryRanges())
+        {
+            if (xMin <= x && x < xMax)
+            {
+                entry = currentEntry;
+                return true;
+            }
+        }
+
+        entry = null;
+        return false;
+    }
+
+    protected override void FrameUpdate(FrameEventArgs args)
+    {
+        // Tween the amounts to their target amounts.
+        const float tweenInverseHalfLife = 8;  // Half life of tween is 1/n
+        bool hasChanged = false;
+
+        foreach (var entry in _entries)
+        {
+            var targetWidthFraction = entry.TargetAmount / Capacity;
+
+            if (entry.WidthFraction == targetWidthFraction)
```

Comment:
don't disable the float equality comment you should be comparing their difference against some epsilon instead of waiting for it to maybe converge on a bit-for-bit equal number

`(entity.WidthFraction - targetWidthFraction) < SOME_EPSILON`

Reply:
This one's actually fine already: The `abs(WidthFraction - targetWidthFraction) < epsilon` is just a couple of lines down. First the function skips any entries that are already exactly the target width, for the rest it calculates the new width and if the new width is close enough to the target width, it sets the width to exactly the target width. Exact comparison is fine here.

I've done a lot of geometry algorithm stuff which is extremely error-prone if you aren't deliberate about your fuzzy float comparisons, so I prefer to be very deliberate with exact & fuzzy comparisons.